### ISSUE TITLE
Desktop app can't be automatically re-opened by the `dx serve`

### ISSUE THREAD

**Issue by Andrew15-5** [DioxusLabs/dioxus#3322]
When re-opening or re-building the app on Linux, it shows:

```
14:43:17 [dev] Build failed: Other(Failed to write main executable

Caused by:
    0: I/O Error: Text file busy (os error 26)
    1: Text file busy (os error 26))
```

**Steps To Reproduce**

- create a Dioxus app for desktop platform
- run `dx serve` on Linux/Unix-like system
- re-open/re-build the app
- see the error

**Expected behavior**

No errors.

**Screenshots**

![image](https://github.com/user-attachments/assets/435fa5ac-f09e-4a0f-972d-d51e6ffe89eb)

**Environment:**
 - Dioxus version: 0.6
 - Rust version: rustc 1.83.0-nightly (fb4aebddd 2024-09-30)
 - OS info: Pop!_OS 22.04 LTS
 - App platform: desktop

**Comment by kingzcheung** [comment]
I'm encountering the same issue.

**Comment by Andrew15-5** [comment]
I've rebased https://github.com/DioxusLabs/dioxus/pull/3323 onto v0.6.1, so you can use it with all the latest updates.

**Comment by jeancf** [comment]
I am seeing the same issue on linux with desktop feature and dx 0.6.1. I am using rustc 1.83.0.

**Comment by Andrew15-5** [comment]
@jeancf, have you tried [the fix](https://github.com/DioxusLabs/dioxus/pull/3323)?

**Comment by jeancf** [comment]
I am happy to try the fix but I don't know how to do it.

**Comment by Andrew15-5** [comment]
```sh
git clone --depth 1 -b fix-desktop-open-linux https://github.com/Andrew15-5/dioxus.git
cd dioxus
cargo install --path ./packages/cli
```

**Comment by jeancf** [comment]
Thanks. I would not have figured it on my own.

The fix indeed closes and restarts the app after a rebuild.

After manually closing the app window, pressing `o` does not reopen it though, but I am not sure this fix was meant to address that as well.

**Comment by crabdancing** [comment]
Can confirm this bug also on NixOS 24.11. Can also confirm that @Andrew15-5's patch fixes the issue. I can _also_ confirm that the app does steal focus when reloaded on my system too, unfortunately. It seems intrinsic problem to re-opening the window. I'm not sure how the tooling behaves on other systems.

I guess if we have a way of somehow re-using the same window, we should do that instead -- otherwise I'd recommend merging this and mitigating the subsequent problem down the line, via either changing window metadata during diagnostic builds to prevent it from raising focus, and try to move it out of the way -- or recommending configs for the Desktop Environment layer (e.g., KDE Window Rules (if the system in question uses KDE). I'm still learning the architecture, so I'm not sure if it is actually possible to simply use the same window.

**Comment by Andrew15-5** [comment]
The app focus issue is honestly the norm, because that's how the windowing system works: you open a new window, it becomes focused. I maybe thought of this once or twice before.

I investigated this for Linux, and there is no way to prevent the focus, unless the app itself can somehow do this:

```sh
id=$(xprop -root _NET_ACTIVE_WINDOW | awk '{print $5}') && ./app & disown && pid=$!; while ! wmctrl -lp | grep "$pid"; do :; done && wmctrl -iR "$id"
```

And AFAIK there is no better way than just wrapping these commands in [`Command`](https://doc.rust-lang.org/std/process/struct.Command.html).

Since https://github.com/jkelleyrtp/ipbp exists, I'm petty sure there is no way to just re-load stuff in-place, since the desktop is literally a single binary.

> force it to move to a more appropriate location on screen.

By default, Dioxus already maintains the same window position on rebuilds, but there is some issue with it: https://github.com/DioxusLabs/dioxus/issues/2318.

> Also, after manually closing the app window, pressing `o` does not reopen it, but I am not sure this fix was meant to address that as well.

This is because currently it tries to find the app from the hashmap and then open it. But, IIUC, the app is being automatically removed from the hashmap because it's being closed and therefore being dropped in `dx`. So this is most likely just an oversight.

I don't know how to properly fix it, because I don't see an immediate way to get the `AppHandle` etc., so instead I just run the re-build action, which in turn opens the built app. Since the rebuilt is a no-op in this case, it takes a tiny bit of additional time to open the app. Here is the fix for this on top of the current fix: https://github.com/Andrew15-5/dioxus/tree/fix-open-button.

**Comment by crabdancing** [comment]
@Andrew15-5 The shell script solution you suggest isn't really one I'd champion, as it doesn't work on Wayland. Wayland protocols for more sophisticated automation are still being developed, and X11 is being dropped by the ecosystem -- so while it might be a temporary fix to have _some_ hook for a script the user can put in, hardcoding in an X11-specific solution seems like a no-go IMO.

This seems out-of-scope for this issue, tho.

**Comment by Andrew15-5** [comment]
Right, I forgot to mention that I only found the solution for the X11, and any tool that can help with that isn't even mandatory, so I'm pretty sure the dev must install them separately. You probably can check the windowing system and run the correct command.

If this can be done for other OSes, then I think this is worth adding. Since there are a lot of unevenness between different OSes as is, I don't think adding this just for X11-based OSes would be that bad. Because better DX even for some group of people (not super small) is better than bad DX for everyone.

I also remember this being annoying, so I would probably use this. And since I already have the patch, I can slap it on top of any version and use it, even if just for myself.

**Comment by jkelleyrtp** [comment]
Just fixed this in #3608

The new system adds a bit of entropy to each .exe name in development that allows us to write multiple variants of the exe in the same build dir while the app is running.

**Comment by Andrew15-5** [comment]
How many generated binaries can be present in the directory at the same time? For the same program. If only 2 then it's fine, but if more, then this is unnecessary bloating as the bare-bones debug binary is over 150 MiB (and 2 of them is already over 300 MiB).

**Comment by jkelleyrtp** [comment]
Only two binaries - we write the new one with the entropy and then clean up the old one
